modules/fwsspBidAdapter.js

```diff
@@ -0,0 +1,716 @@
+    }
+
+    if (gppConsent) {
+      if (typeof gppConsent.gppString === 'string') {
+        params.gpp = gppConsent.gppString;
+      }
+      if (gppConsent.applicableSections) {
+        params.gpp_sid = gppConsent.applicableSections;
+      }
+    }
+
+    let queryString = '';
+    if (params) {
+      queryString = '?' + `${formatQS(params)}`;
+    }
+
+    const syncs = [];
+    if (syncOptions && syncOptions.pixelEnabled) {
+      syncs.push({
+        type: 'image',
+        url: USER_SYNC_URL + queryString
+      });
+    } else if (syncOptions.iframeEnabled) {
+      syncs.push({
+        type: 'iframe',
+        url: USER_SYNC_URL + queryString
+      });
+    }
+
+    return syncs;
+  }
+}
+
+/**
+ * Generates structured HTML for FreeWheel MRM ad integration with Prebid.js
+ * @param {Object} bidrequest - Prebid bid request
+ * @param {number[]} size - Prebid ad dimensions [width, height]
+ * @returns {string} Formatted HTML string for ad rendering
+ */
+export function formatAdHTML(bidrequest, size) {
+  const sdkUrl = getSdkUrl(bidrequest);
+  const displayBaseId = 'fwssp_display_base';
+
+  const startMuted = typeof bidrequest.params.isMuted == 'boolean' ? bidrequest.params.isMuted : true
+  const showMuteButton = typeof bidrequest.params.showMuteButton == 'boolean' ? bidrequest.params.showMuteButton : false
+
+  let playerParams = null;
+  try {
+    playerParams = JSON.stringify(bidrequest.params.playerParams);
+  } catch (error) {
+    logWarn('Error parsing playerParams:', error);
+  }
+
+  return `<div id='${displayBaseId}' class='ad-container' style='width:${size[0]}px;height:${size[1]}px;'>
+  <script type='text/javascript'>
+    const script = document.createElement('script');
+    script.src = '${sdkUrl}';
+    script.async = true;
+
+    const topWindow = function() {
+      let res = window;
+      try {
+        while (top !== res) {
+          if (res.parent.location.href.length) {
+            res = res.parent;
+          }
+        }
+      } catch (e) {}
+      return res;
+    }();
+    const cache = topWindow.fwssp_cache ? topWindow.fwssp_cache['${bidrequest.adUnitCode}'] : null;
+    const vastResponse = cache ? cache.response : null;
+    const listeners = cache ? cache.listeners : null;
+
+    const config = {
+      displayBaseId: '${displayBaseId}',
+      vastResponse: vastResponse,
+      showMuteButton: ${showMuteButton},
+      startMuted: ${startMuted},
+      playerParams: ${playerParams},
+      format: 'outstream',
+      listeners: listeners
+    };
+
+    const timeoutId = setTimeout(() => {
+      console.warn('MRM AdManager load timeout: 5000 ms');
+      cleanup();
+    }, 5000);
+
+    script.onload = () => {
+      clearTimeout(timeoutId);
+      this.prebidPlayer = window.tv.freewheel.SDK.newPrebidPlayer(config);
+    };
+
+    script.onerror = (err) => {
+      cleanup();
+      console.warn('MRM AdManager load failed:', err.message);
+    };
+
+    cleanup = () => {
```

Comment:
Global scope, how about 

(function () {
    const cleanup = () => {
      /* cleanup logic */
    };

    // remainder of the ad setup...
  })();